### PR TITLE
ARTP-250 Add styleguide

### PR DESCRIPTION
**Pull request by evanrs**
### Desktop
styleguide 1040

### Mobile
styleguide iphone x

### CODE REVIEW

File: src/client/.prettierrc [hunk 1]
```diff
@@ -1,5 +1,6 @@
 {
   "semi": false,
   "singleQuote": true,
-  "printWidth": 120
+  "printWidth": 120,
+  "trailingComma": "all"
```
Comment (keshaldesai) [inline src/client/.prettierrc]: Changes like this should really be in their own PR if possible.

File: src/client/package.json [hunk 2]
```diff
@@ -54,6 +54,12 @@
       "git add"
     ]
   },
+  "prettier": {
+    "semi": false,
+    "singleQuote": true,
+    "printWidth": 120,
+    "trailingComma": "all"
+  },
```
Comment (keshaldesai) [inline src/client/package.json]: Why does the prettier config also ned to be here?

File: src/client/src/routes/StyleguideRoute/StyleguideRoute.tsx [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    ColorSpectrum,
+    CoreBranding,
+    FontFamilies,
+    Atoms,
+    Ending: () => <SectionBlock mh={5} intent="inverted" />,
+  },
+  (value, key) => _.kebabCase(key),
+)
+
+export class StyleguideRoute extends React.Component {
```
Comment (keshaldesai) [inline src/client/src/routes/StyleguideRoute/StyleguideRoute.tsx]: If you wanted, this looks like it could become a functional component.

File: src/client/src/routes/StyleguideRoute/StyleguideRoute.tsx [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+import FloatingTools from './components/FloatingsTools'
+import { Block, SectionBlock } from './styled'
+
+import Atoms from './sections/Atoms'
+import ColorSpectrum from './sections/ColorSpectrum'
+import CoreBranding from './sections/CoreBranding'
+import FontFamilies from './sections/FontFamilies'
+import Introduction from './sections/Introduction'
+
+const sections = _.mapKeys(
```
Comment (keshaldesai) [inline src/client/src/routes/StyleguideRoute/StyleguideRoute.tsx]: There is a lot of dependency on lodash in this file. Are there any areas where it might be excessive?

File: src/client/src/routes/StyleguideRoute/components/ButtonGrid.tsx [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+import _ from 'lodash'
+import React from 'react'
+
+import { Button, ButtonGroup } from 'rt-styleguide'
+import { resolvesColor, styled } from 'rt-theme'
+
+import { PassThroughProps } from '../tools'
+
+export default class ButtonGrid extends React.Component {
+  state = {}
```
Comment (keshaldesai) [inline src/client/src/routes/StyleguideRoute/components/ButtonGrid.tsx]: Remove

File: src/client/src/routes/StyleguideRoute/components/ButtonGrid.tsx [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
+import _ from 'lodash'
+import React from 'react'
+
+import { Button, ButtonGroup } from 'rt-styleguide'
+import { resolvesColor, styled } from 'rt-theme'
+
+import { PassThroughProps } from '../tools'
+
+export default class ButtonGrid extends React.Component {
```
Comment (keshaldesai) [inline src/client/src/routes/StyleguideRoute/components/ButtonGrid.tsx]: Can be a functional component.

File: src/client/src/routes/StyleguideRoute/components/ColorBlocks.tsx [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+import { PassThroughProps } from '../tools'
+
+const { spectrum } = colors
+
+const SWATCH_KEYS = {
+  light: [95, 9, 8, 7, 6, 5, 4, 3, 2, 1].map(v => `L${v}`),
+  dark: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10].map(v => `D${v}`),
+}
+
+export interface ColorBlocksProps extends PassThroughProps {}
```
Comment (keshaldesai) [inline src/client/src/routes/StyleguideRoute/components/ColorBlocks.tsx]: If there is nothing added to the extended interface, why extend?

File: src/client/src/routes/StyleguideRoute/components/ColorBlocks.tsx [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+
+export const ColorBlocks: React.SFC<ColorBlocksProps> = props => {
```
Comment (keshaldesai) [inline src/client/src/routes/StyleguideRoute/components/ColorBlocks.tsx]: Can be functional.

File: src/client/src/routes/StyleguideRoute/elements.tsx [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+import { resolvesColor } from 'rt-theme'
+
+import { mapMarginPaddingProps, mapTextProps, MarginPaddingProps, TextProps } from './styled'
+
+export { Paragraph } from './styled'
+
+export const Hashable = ({ is: Element = 'div', ...props }) => <Element {...props} id={_.kebabCase(props.children)} />
+
+interface HeaderProps extends MarginPaddingProps, TextProps {}
+export const Header: Styled<HeaderProps> = styled(Hashable)`
```
Comment (keshaldesai) [inline src/client/src/routes/StyleguideRoute/elements.tsx]: General comment for all the places you have this:

`Component: Styled<Props>` --> `Component = styled<Props>('div')`

You shouldn't need the `Styled` interface.

File: src/client/src/routes/StyleguideRoute/mapProps.tsx [hunk 10]
```diff
@@ -0,0 +1,115 @@
+/**
+ * Copied from @evanrs/map-props
+ * https://github.com/evanrs/map-props
+ */
```
Comment (keshaldesai) [inline src/client/src/routes/StyleguideRoute/mapProps.tsx]: I don't think this is appropriate.

File: src/client/src/routes/StyleguideRoute/styled/Paragraph.tsx [hunk 11]
```diff
@@ -0,0 +1,17 @@
+import { styled, Styled } from 'rt-theme'
+
+import { mapMarginPaddingProps, MarginPaddingProps } from './mapMarginPaddingProps'
+import { mapTextProps, TextProps } from './Text'
+
+export interface ParagraphProps extends TextProps, MarginPaddingProps {}
```
Comment (keshaldesai) [inline src/client/src/routes/StyleguideRoute/styled/Paragraph.tsx]: `type ParagraphProps = TextProps & MarginPaddingProps`

File: src/client/src/routes/StyleguideRoute/styled/SectionBlock.tsx [hunk 12]
```diff
@@ -0,0 +1,125 @@
+import _ from 'lodash'
+import React from 'react'
+import { css, styled, Styled } from 'rt-theme'
+
+import { Block, BlockProps } from '../styled'
+import { mapMarginPaddingProps, MarginPaddingProps } from './mapMarginPaddingProps'
+
+export interface SectionProps extends BlockProps, MarginPaddingProps {
+  mh?: any
```
Comment (keshaldesai) [inline src/client/src/routes/StyleguideRoute/styled/SectionBlock.tsx]: Is this really an `any`?

File: src/client/src/routes/StyleguideRoute/styled/mapMarginPaddingProps.ts [hunk 13]
```diff
@@ -0,0 +1,95 @@
+import _ from 'lodash'
+import { css } from 'rt-theme'
+
+import { mapProps } from '../tools'
+
+// export type PropValue = number | string | boolean
+export type PropValue = any
```
Comment (keshaldesai) [inline src/client/src/routes/StyleguideRoute/styled/mapMarginPaddingProps.ts]: You should be able to type this.

File: src/client/src/routes/StyleguideRoute/styled/mapMarginPaddingProps.ts [hunk 14]
```diff
@@ -0,0 +1,95 @@
...
+  py?: PropValue
+  pt?: PropValue
+  pr?: PropValue
+  pb?: PropValue
+  pl?: PropValue
+}
+
+export interface MarginPaddingProps extends MarginProps, PaddingProps {}
+
+export const { marginPaddingProps, mapMarginPaddingProps } = (() => {
```
Comment (keshaldesai) [inline src/client/src/routes/StyleguideRoute/styled/mapMarginPaddingProps.ts]: Some comments might help the user understand what you're trying to accomplish.

File: src/client/src/routes/StyleguideRoute/styled/rules.ts [hunk 15]
```diff
@@ -0,0 +1,44 @@
+import { css } from 'rt-theme'
+
+export { mapMarginPaddingProps, MarginPaddingProps } from './mapMarginPaddingProps'
+
+export const preventStutter = css`
+  -webkit-backface-visibility: hidden;
```
Comment (keshaldesai) [inline src/client/src/routes/StyleguideRoute/styled/rules.ts]: I think that even though these properties are unsupported or experimental, the preprocessor will auto-prefix.

File: src/client/src/rt-styleguide/components/Button.tsx [hunk 16]
```diff
@@ -111,7 +111,7 @@ export class ButtonGroup extends React.Component<ButtonStyleProps> {
   }
 }
 
-const BaseElement = styled.button<ButtonStyleProps>`
+const StyledBase = withProps({ role: 'button' })(styled.div`
```
Comment (keshaldesai) [inline src/client/src/rt-styleguide/components/Button.tsx]: Are we adding `recompose` back just for this one case?

File: src/client/src/rt-theme/globals.ts [hunk 17]
```diff
@@ -53,24 +53,21 @@ export default injectGlobal`
...
-  ::-webkit-scrollbar-corner {
-    background-color: rgba(0,0,0,0);
+    @media all and (max-width: 400px) {
+      font-size: 14px;
+    }
+
+    @media all and (max-width: 320px) {
+      font-size: 12px;
+    }
+  }
```
Comment (keshaldesai) [inline src/client/src/rt-theme/globals.ts]: 👍

File: src/client/src/rt-theme/index.ts [hunk 18]
```diff
@@ -2,9 +2,9 @@ export { flush, hydrate, cx, merge, getRegisteredStyles, injectGlobal, keyframes
...
 
-export { colors } from './colors'
+export { colors, CorePalette, CorePaletteMap, ColorPalette, ColorPaletteMaps } from './colors'
 export { themes, Theme } from './themes'
-export { resolvesColor } from './tools'
+export { resolvesColor, Selector } from './tools'
 export { ThemeProvider } from './ThemeProvider'
 export { ThemeState, ThemeStateValue } from './ThemeState'
-export { default as styled } from './styled'
+export { default as styled, Styled, StyledProps } from './styled'
```
Comment (keshaldesai) [inline src/client/src/rt-theme/index.ts]: You shouldn't need `Styled` or `StyledProps`.

File: src/client/src/shell/routes/ShellRoute.tsx [hunk 19]
```diff
@@ -78,7 +78,7 @@ const WorkspaceWrapper = styled(Wrapper)`
 `
 
 const AnalyticsWrapper = styled(Wrapper)`
-  padding: 0.375rem 1.25rem 0 0;
+  padding: 0.5rem 0.5rem;
```
Comment (keshaldesai) [inline src/client/src/shell/routes/ShellRoute.tsx]: `padding: 0.5rem`

File: src/client/src/ui/analytics/components/pnlChart/chartGradient.ts [hunk 20]
```diff
@@ -11,9 +11,9 @@ import 'classlist-polyfill'
  * read about the SVG defs:
  * https://developer.mozilla.org/en-US/docs/Web/SVG/Element/defs
  *
- * There are four colour stops: top, bottom, and the two in the middle that have the same offset value
+ * There are four color stops: top, bottom, and the two in the middle that have the same offset value
```
Comment (keshaldesai) [inline src/client/src/ui/analytics/components/pnlChart/chartGradient.ts]: 😉

File: src/client/.prettierrc (hunk 1 above)
Comment (evanrs) [reply src/client/.prettierrc]: I agree — definitely an oversight.

File: src/client/package.json (hunk 2 above)
Comment (evanrs) [reply src/client/package.json]: Paranoia involving VSCode — I'll remove.

File: src/client/src/routes/StyleguideRoute/StyleguideRoute.tsx (hunk 4 above)
Comment (evanrs) [reply src/client/src/routes/StyleguideRoute/StyleguideRoute.tsx]: There is no meaningful cost due to the use of `babel-lodash-plugin` — and no, I don't think use of a functional tool like this is excessive.

Additionally, the tooling around `lodash` is some of the best — therefore it shouldn't impose any costs on the developer. And, while the API should be obvious simply by convention it also comes with complete documentation in VSCode or Atom with an extension.

There is every reason to use it where it provides a relevant functional primitive that simplifies the expression and adds no cognitive or runtime overhead.

File: src/client/src/routes/StyleguideRoute/components/ButtonGrid.tsx (hunk 5 above)
Comment (evanrs) [reply src/client/src/routes/StyleguideRoute/components/ButtonGrid.tsx]: 👍

File: src/client/src/rt-styleguide/components/Button.tsx (hunk 16 above)
Comment (evanrs) [reply src/client/src/rt-styleguide/components/Button.tsx]: I think the right framing is with consideration that `Emotion` is an  incomplete implementation of `styled-components`. While `styled-components` implements the `attrs` method to address this very common use-case, the author of `Emotion` refuses to — instead they prescribe this approach. As the officially documented approach to an existing and expected API I think we should resist introducing additional indirection or variation on a variation.

I will look to see if I can remove it simply by adding it to the wrapping components — as this component is not exported. However, where similar cases emerge I believe this framing is the correct approach to such an issue.

File: src/client/src/rt-theme/index.ts (hunk 18 above)
Comment (evanrs) [reply src/client/src/rt-theme/index.ts]: There are two points to address:
1. Some APIs entirely fail when employing the `styled.div<props>` approach. I would encourage anyone finding a solution.
2. This approach is consistent with the typing of any other non-class component
```ts
const SomeComponent: React.SFC<SomeComponentProps> = () => 
```

File: src/client/src/shell/routes/ShellRoute.tsx (hunk 19 above)
Comment (evanrs) [reply src/client/src/shell/routes/ShellRoute.tsx]: Fair.

File: src/client/src/routes/StyleguideRoute/StyleguideRoute.tsx (hunk 3 above)
Comment (evanrs) [reply src/client/src/routes/StyleguideRoute/StyleguideRoute.tsx]: 👍

File: src/client/src/routes/StyleguideRoute/components/ColorBlocks.tsx (hunk 8 above)
Comment (evanrs) [reply src/client/src/routes/StyleguideRoute/components/ColorBlocks.tsx]: Do you mean it can be implicit return?

File: src/client/src/routes/StyleguideRoute/elements.tsx (hunk 9 above)
Comment (evanrs) [reply src/client/src/routes/StyleguideRoute/elements.tsx]: It continues to be a useful tool — at times typescript objects to that method. Additionally the typescript + styled tooling is incomplete and features are lost by that approach. Correction and formatting on CSS is lost, for example. And, as mentioned before, it is actually less consistent without how most types are described.

File: src/client/src/routes/StyleguideRoute/mapProps.tsx (hunk 10 above)
Comment (evanrs) [reply src/client/src/routes/StyleguideRoute/mapProps.tsx]: How so? It would be appropriate without the comment?

File: src/client/src/routes/StyleguideRoute/elements.tsx (hunk 9 above)
Comment (evanrs) [reply src/client/src/routes/StyleguideRoute/elements.tsx]: How would you achieve typing on the `is` prop — a convention from `styled–system`:
https://github.com/AdaptiveConsulting/ReactiveTraderCloud/pull/854#diff-612387fc2db9d8ac0322d7125d4ea86cR137

File: src/client/src/routes/StyleguideRoute/mapProps.tsx (hunk 10 above)
Comment (evanrs) [reply src/client/src/routes/StyleguideRoute/mapProps.tsx]: This specific file only serves `mapMarginPaddingProps` a useful tool for design heavy marketing content — this mimics `styled-system` convention of `px …`, `mx …` for adding white space. Unfortunately Styled System does not support Emotion — otherwise I'd readily advise adopting it here.

And, because the conventions like this rest on a great deal of existing work and experience it will necessarily not be a wholly unique implementation. However, as this code now resides within the code base it poses issue of governance or maintenance that might present a risk to Reactive Trader in the long term.

Comment (evanrs) [reply src/client/src/routes/StyleguideRoute/mapProps.tsx]: However, `mapMarginPaddingProps` can be rewritten to isolate the implementation there.

File: src/client/src/routes/StyleguideRoute/styled/Paragraph.tsx (hunk 11 above)
Comment (evanrs) [reply src/client/src/routes/StyleguideRoute/styled/Paragraph.tsx]: But, is there any reason to not describe it as `interface`? For example, can a `type` be extended by another `interface`?

File: src/client/src/routes/StyleguideRoute/mapProps.tsx (hunk 10 above)
Comment (keshaldesai) [reply src/client/src/routes/StyleguideRoute/mapProps.tsx]: It is my opinion that we should not be tagging our work on RT. If you're comfortable with leaving it here without your name, I think that's fine. However, if you're not, I would suggest rewriting or finding a replacement. Again, this is my personal opinion and I don't mind if you find someone to overrule this.

File: src/client/src/routes/StyleguideRoute/styled/mapMarginPaddingProps.ts (hunk 14 above)
Comment (evanrs) [reply src/client/src/routes/StyleguideRoute/styled/mapMarginPaddingProps.ts]: I will add a comment at the top to this documentation — in general anything that gets closer to `styled-system` for this kind of primitive is ideal.

https://jxnblk.com/styled-system/getting-started#margin--padding

File: src/client/src/routes/StyleguideRoute/elements.tsx (hunk 9 above)
Comment (keshaldesai) [reply src/client/src/routes/StyleguideRoute/elements.tsx]: Leave a comment on the exports of where you create the `Styled` type to explain why you're using them.

File: src/client/src/rt-theme/index.ts (hunk 18 above)
Comment (keshaldesai) [reply src/client/src/rt-theme/index.ts]: 1. That makes sense.
2. TypeScript is able to infer from the styled function. In the example you provide, TypeScript cannot infer that this function is specifically `React.SFC`, hence the explicit type.
